src/libsystemd/sd-bus/bus-message.c

```diff
@@ -5833,34 +5835,37 @@ int bus_message_remarshal(sd_bus *bus, sd_bus_message **m) {
         if ((*m)->destination && !n->destination) {
                 r = message_append_field_string(n, BUS_MESSAGE_HEADER_DESTINATION, SD_BUS_TYPE_STRING, (*m)->destination, &n->destination);
                 if (r < 0)
-                        return r;
+                        goto fail;
         }
 
         if ((*m)->sender && !n->sender) {
                 r = message_append_field_string(n, BUS_MESSAGE_HEADER_SENDER, SD_BUS_TYPE_STRING, (*m)->sender, &n->sender);
                 if (r < 0)
-                        return r;
+                        goto fail;
         }
 
         n->header->flags |= (*m)->header->flags & (BUS_MESSAGE_NO_REPLY_EXPECTED|BUS_MESSAGE_NO_AUTO_START);
 
         r = sd_bus_message_copy(n, *m, true);
         if (r < 0)
-                return r;
+                goto fail;
 
         timeout = (*m)->timeout;
         if (timeout == 0 && !((*m)->header->flags & BUS_MESSAGE_NO_REPLY_EXPECTED))
                 timeout = BUS_DEFAULT_TIMEOUT;
 
         r = sd_bus_message_seal(n, BUS_MESSAGE_COOKIE(*m), timeout);
         if (r < 0)
-                return r;
+                goto fail;
 
         sd_bus_message_unref(*m);
         *m = n;
         n = NULL;
 
         return 0;
+fail:
+        sd_bus_message_unref(n);
```

Comment:
There is `_cleanup_(sd_bus_message_unrefp) sd_bus_message *n` at the top of the function, which should be enough to prevent any leaks. And it seems that additional call to `sd_bus_message_unref` should trigger `assert`, because `sd_bus_message_unref` is not supposed to be called two times in a row.

Reply:
Could you show a way to reproduce the leak you are fixing? It would also be nice if you could attach the output of `valgrind`, `asan` or something similar.

Reply:
Ah, I totally missed it. Will remove that part. I found this during code review not via memory leak check tool.